test/operance_data_controller_test.dart

```diff
@@ -0,0 +1,444 @@
+        expect(controller.allRows, isEmpty);
+      });
+    });
+
+    group('When fetching data with empty pages and no rows returned', () {
+      test('Then it should add an empty list to pages', () async {
+        await initializeController(
+          initialPage: emptyPage,
+        );
+
+        await controller.nextPage();
+
+        expect(controller.currentRows, isEmpty);
+        expect(controller.allRows, isEmpty);
+        expect(controller.pages.length, 1);
+        expect(controller.pages.first, isEmpty);
+      });
+    });
+
+    group('When initialized with default values', () {
+      test('Then it should have default values', () {
+        expect(controller.columnOrder, isEmpty);
+        expect(controller.hiddenColumns, isEmpty);
+        expect(controller.currentRows, isEmpty);
+        expect(controller.searchedRows, isEmpty);
+        expect(controller.selectedRows, isEmpty);
+        expect(controller.expandedRows, isEmpty);
+        expect(controller.sorts, isEmpty);
+        expect(controller.rowsPerPage, rowsPerPage);
+        expect(controller.currentPageIndex, 0);
+        expect(controller.isLoading, isFalse);
+        expect(controller.hoveredRowIndex, isNull);
+        expect(controller.canGoNext, isFalse);
+        expect(controller.canFetchNext, isFalse);
+        expect(controller.canGoPrevious, isFalse);
+      });
+    });
+
+    group('When initialized with parameters', () {
+      test('Then it should set the provided values', () async {
+        await initializeController(
+          columnOrder: columnOrder,
+          initialPage: initialPageWithRows,
+          currentPageIndex: currentPageIndex,
+          rowsPerPage: rowsPerPage10,
+          infiniteScroll: true,
+        );
+
+        expect(controller.columnOrder, columnOrder);
+        expect(controller.currentRows, <String>['Row1', 'Row2']);
+        expect(controller.rowsPerPage, rowsPerPage10);
+        expect(controller.currentPageIndex, currentPageIndex);
+        expect(controller.canGoNext, isFalse);
+        expect(controller.canFetchNext, isFalse);
+        expect(controller.canGoPrevious, isTrue);
+      });
+    });
+
+    group('When navigating pages', () {
+      setUp(() async {
+        await initializeController(
+          initialPage: initialPage,
+          rowsPerPage: rowsPerPage10,
+        );
+      });
+
+      test('Then it should navigate to the next page', () async {
+        await controller.nextPage();
+
+        expect(controller.currentPageIndex, currentPageIndex);
+        expect(
+          controller.currentRows,
+          initialPage.$1.sublist(10, 20),
+        );
+      });
+
+      test('Then it should navigate to the previous page', () async {
+        await controller.nextPage();
+
+        controller.previousPage();
+
+        expect(controller.currentPageIndex, 0);
+        expect(controller.currentRows, initialPage.$1.sublist(0, 10));
+      });
+    });
+
+    group('When sorting columns', () {
```

Comment:
These could be more comprehensive. Maybe add tests for the following:

Multiple column sorting
Sort stability
Sort with null values
Custom sort comparators